File: backend/app/schemas/ws.py
"""WebSocket message schemas (for documentation + validation)."""
from __future__ import annotations

from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict


# ─── Public orderbook channel ────────────────────────────────────────────────

class OrderBookSnapshot(BaseModel):
    """Full L2 snapshot sent on initial subscription."""
    event: Literal["orderbook_snapshot"] = "orderbook_snapshot"
    symbol: str
    bids: list[list[Decimal]]  # [[price, volume], ...]
    asks: list[list[Decimal]]
    last_trade_price: Decimal | None = None
    ts: int


class OrderBookUpdate(BaseModel):
    """Incremental L2 update (delta)."""
    event: Literal["orderbook_update"] = "orderbook_update"
    symbol: str
    changes: list[dict]  # [{"side": "bid"|"ask", "price": ..., "qty": ...}]
    ts: int


class TradePrint(BaseModel):
    """A single trade print broadcast on the public trades channel."""
    event: Literal["trade"] = "trade"
    symbol: str
    trade_id: int
    price: Decimal
    quantity: Decimal
    side: str  # taker side
    ts: int


class WSPing(BaseModel):
    """Heartbeat ping."""
    event: Literal["ping"] = "ping"
    ts: int


# ─── Private channel ─────────────────────────────────────────────────────────

class OrderUpdate(BaseModel):
    """Per-user order status update."""
    event: Literal["order"] = "order"
    order_id: int
    symbol: str
    side: str
    type: str
    status: str
    status_event: str  # placed | partially_filled | filled | canceled | modified | rejected | triggered
    price: Decimal | None = None
    quantity: Decimal | None = None
    filled_quantity: Decimal | None = None
    remaining_quantity: Decimal | None = None
    avg_fill_price: Decimal | None = None
    last_trade_qty: Decimal | None = None
    last_trade_price: Decimal | None = None
    client_order_id: str | None = None
    bulk_id: str | None = None
    ts: int


class BalanceUpdate(BaseModel):
    """Per-user balance update."""
    event: Literal["balance"] = "balance"
    asset: str
    total: Decimal
    locked: Decimal
    available: Decimal
    change: Decimal | None = None
    reason: str | None = None
    order_id: int | None = None
    ts: int


class BulkResult(BaseModel):
    """Bulk operation result notification."""
    event: Literal["bulk_result"] = "bulk_result"
    bulk_id: str
    action: str
    total: int
    succeeded: int
    failed: list[dict]
    ts: int


class SLTPActivated(BaseModel):
    """Notification that SL/TP children have been activated."""
    event: Literal["sl_tp_activated"] = "sl_tp_activated"
    parent_order_id: int
    sl_order_id: int | None = None
    tp_order_id: int | None = None
    ts: int


# ─── Subscribe / auth messages ───────────────────────────────────────────────

class WSSubscribe(BaseModel):
    """Client → server subscription message."""
    action: Literal["subscribe"]
    channel: str
    symbol: str | None = None
    depth: int = 20


class WSAuth(BaseModel):
    """Client → server auth message (for private channel)."""
    action: Literal["auth"]
    api_key: str
    timestamp: int
    signature: str


__all__ = [
    "OrderBookSnapshot",
    "OrderBookUpdate",
    "TradePrint",
    "WSPing",
    "OrderUpdate",
    "BalanceUpdate",
    "BulkResult",
    "SLTPActivated",
    "WSSubscribe",
    "WSAuth",
]

File: frontend/frontend/src/types/ws.ts
// WebSocket message types (matching backend schemas/ws.py)

export type WSEventType =
  | "orderbook_snapshot"
  | "orderbook_update"
  | "trade"
  | "order"
  | "balance"
  | "bulk_result"
  | "sl_tp_activated"
  | "ping"
  | "pong"
  | "auth_required"
  | "auth_ok"
  | "auth_failed"
  | "auth_timeout";

export interface WSMessage {
  event: WSEventType;
  ts: number;
  [key: string]: unknown;
}

export interface OrderBookSnapshot extends WSMessage {
  event: "orderbook_snapshot";
  symbol: string;
  bids: [number, number][]; // [price, volume]
  asks: [number, number][];
  last_trade_price: number | null;
}

export interface OrderBookChange {
  side: "bid" | "ask";
  price: number;
  qty: number;
}

export interface OrderBookUpdate extends WSMessage {
  event: "orderbook_update";
  symbol: string;
  changes: OrderBookChange[];
}

export interface TradePrint extends WSMessage {
  event: "trade";
  symbol: string;
  trade_id: number;
  price: number;
  quantity: number;
  side: string;
}

export interface OrderUpdate extends WSMessage {
  event: "order";
  order_id: number;
  symbol: string;
  side: string;
  type: string;
  status: string;
  status_event: string;
  price?: number;
  quantity?: number;
  filled_quantity?: number;
  remaining_quantity?: number;
  avg_fill_price?: number;
  client_order_id?: string;
  bulk_id?: string;
}

export interface BalanceUpdate extends WSMessage {
  event: "balance";
  asset: string;
  total: number;
  locked: number;
  available: number;
  change?: number;
  reason?: string;
  order_id?: number;
}

export interface BulkResult extends WSMessage {
  event: "bulk_result";
  bulk_id: string;
  action: string;
  total: number;
  succeeded: number;
  failed: { index: number; code: string; message: string }[];
}
